File: leetcode/Algorithms/152.MaximumProductSubarray/Solution.java
public class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0], a = 1, b = 1;

        for (int num : nums) {
            result = Math.max(result, a *= num);
            if (num == 0) a = 1;
        }

        for (int i = nums.length - 1; i >= 0; i--) {
            result = Math.max(result, b *= nums[i]);
            if (nums[i] == 0) b = 1;
        }

        return result;
    }
}

File: leetcode/Algorithms/152.MaximumProductSubarray/Solution.go
func maxProduct(nums []int) int {
	result, max, min := nums[0], nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] < 0 {
			min, max = max, min
		}
		min = getMin(nums[i], nums[i]*min)
		max = getMax(nums[i], nums[i]*max)
		result = getMax(result, max)
	}
	return result
}

func getMax(a int, b int) int {
	if a < b {
		return b
	}
	return a
}

func getMin(a int, b int) int {
	if a < b {
		return a
	}
	return b
}
